proposals/0281-visible-forall.rst

```diff
@@ -570,6 +614,12 @@ information has been fully determined:
   1. Namespace selection syntax (under ``-XDataKinds``)
   2. Name quotation syntax (under ``-XTemplateHaskell``)
 
+* The meaning of ``->`` has been determined. It can stand for one of the following:
+
+  1. Part of lambda, case-of, or multi-if syntax, as in ``\x -> ...``
+  2. The function type constructor ``(->)``, as in ``Int -> Bool``
+  3. A view pattern ``f (e -> p) = ...`` (under ``-XViewPatterns``)
+
```

Comment:
> We need to say "Extend the syntax of both terms and patterns with several constructs that..."

The dichotomies I use throughout the proposal is (term-vs-type)-syntax and expression-vs-pattern. So I can't say "terms and patterns", that'd be a category error.

I edited the sentence to say "extend the term syntax (expressions and patterns)".

> Then the little table with "proposed" vs "as opposed to" needs to be duplicated for patterns. Or maybe add a third column.

I don't think that's necessary. GHC uses the same grammar for expressions and patterns. We can pretend otherwise when it's convenient, but here it's more convenient to acknowledge this and avoid the duplication of nonterminals.

I added the following sentences: "Patterns reuse the expression grammar in GHC. The new "terms-in-types" productions are also valid in patterns."

> Then new line 653 "This is a necessity..." What is "this"? Please clarify.

Clarified as follows: "In `e1 -> e2` and `e1 => e2`, using the expression grammar for the LHS is a necessity to avoid parsing conflicts"

> Then the following (i), (ii), (iii) is a mystery. We have extended the syntax of terms. Very well. If we later want to re-interpret that term as a type, we have work to do (Section 7.6). But picking out three particular cases here seems quite confusing. What is special about them? It's all handled in 7.6 isn't it?

Nothing special.  Moved these examples to "Effect and Interactions".

> Can we just delete this cryptic utterance, whose meaning is unclear, and instead spell it out where needed?

OK, done.

> What is this "interpret" business? 

Well, the production is always there. The GHC parser accepts `p1 -> p2` today! It just interprets it as a view pattern, not as a function arrow. I'm using "interprets" not in the sense of an interpreter or compiler, but in the sense of assigning meaning to syntax (in the same way a person interprets what they read or hear). 

Anyway, I rephrased it in a way that better aligns with your suggestion.

> `case ... of x -> (y -> z)` cannot possibly make sense, right?

Not with this proposal alone, no. But it will make sense if we allow to return types from functions (it's far-fetched but I like to think it'll be possible in DH)

```haskell
f :: Bool -> Type
f b = case b of
  True  -> Int
  False -> (Int -> String)
```

It's not something I'm proposing here, but it's in the back of my mind when I'm designing the syntax. I also expect to be able to match on types

```haskell
g :: Type -> Bool
g (Int -> String) = True
g _ = False
```

Both examples above are already possible with `TypeRep`, of course, but the syntax is nowhere as nice.

Reply:
> I don't think that's necessary. GHC uses the same grammar for expressions and patterns. We can pretend otherwise when it's convenient, but here it's more convenient to acknowledge this and avoid the duplication of nonterminals.

But GHC's implementation is, well, just an implementation. The Haskell Report uses separate syntax for each.  I really think it would be better to be explicit, especially since (p1 -> p2) is treated differently to (e1 -> e2).

I really think it'd be better to be explicit about this.  It's a very confusing area, as we have discovered through our extended interaction.  The more explicit the better.  A bit of duplication does no harm.

Reply:
I added a proper BNF. Having done that, I agree it's an improvement.

Reply:
Thanks! 

Can you add a pointer (in a comment) to point (4) to the production for `(->)` for patterns?

It looks as if these are these give the *complete* syntax for `exp` and `pat`.  Is that right?  Doesn't seem plausible. What did the previous syntax for `exp` and `pat` look like?

Reply:
I pushed another update. Took me longer than expected because I've discovered a precedence issue and had to come up with a mitigation plan

```
f (e -> p :: t)   = ...     -- user-written function LHS
f (e -> (p :: t)) = ...     -- old parse
f ((e -> p) :: t) = ...     -- new parse
```

See the new commit.